From 0e623311ad867bc1a6af443609fb96d0a2380bb7 Mon Sep 17 00:00:00 2001
From: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Date: Mon, 17 Oct 2022 11:31:52 -0700
Subject: [PATCH 7/7] ath12k: Avoid too many dereference to find the rbm id

Use a local structure to do the lookup and find the rbm id.
Avoid too many dereference as it increases the instructions and
brings down the throughput by 400Mbps

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index d71559a..4b14059 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -127,6 +127,26 @@ static void ath12k_hal_tx_cmd_ext_desc_setup(struct ath12k_base *ab, void *cmd,
 					   ti->encrypt_type);
 }
 
+static const struct ath12k_hal_tcl_to_wbm_rbm_map
+ath12k_hal_qcn92xx_tcl_to_wbm_rbm_map[DP_TCL_NUM_RING_MAX] = {
+	{
+		.wbm_ring_num = 0,
+		.rbm_id = HAL_RX_BUF_RBM_SW0_BM,
+	},
+	{
+		.wbm_ring_num = 1,
+		.rbm_id = HAL_RX_BUF_RBM_SW1_BM,
+	},
+	{
+		.wbm_ring_num = 2,
+		.rbm_id = HAL_RX_BUF_RBM_SW2_BM,
+	},
+	{
+		.wbm_ring_num = 4,
+		.rbm_id = HAL_RX_BUF_RBM_SW4_BM,
+	}
+};
+
 int ath12k_dp_tx_direct(struct ath12k_vif *arvif, struct sk_buff *skb)
 {
 	struct ath12k *ar = arvif->ar;
@@ -143,7 +163,7 @@ int ath12k_dp_tx_direct(struct ath12k_vif *arvif, struct sk_buff *skb)
 
 	ring_id = smp_processor_id();
 
-	rbm_id = ab->hal.ops->tcl_to_wbm_rbm_map[ring_id].rbm_id;
+	rbm_id = ath12k_hal_qcn92xx_tcl_to_wbm_rbm_map[ring_id].rbm_id;
 
 	tx_ring = &dp->tx_ring[ring_id];
 
-- 
2.17.1

